Sparse vectors API and local mode

agourlay opened the pull request:
This PR adds support for sparse vectors with:
- local mode
- REST API
- gRPC API (incomplete)

agourlay commented on tools/generate_rest_client.sh:
```diff
@@ -6,7 +6,7 @@ PROJECT_ROOT="$(pwd)/$(dirname "$0")/../"
 cd $(mktemp -d)
 
-git clone --sparse --filter=blob:none --depth=1 -b dev git@github.com:qdrant/qdrant.git
+git clone --sparse --filter=blob:none --depth=1 -b sparse-index-try-to-extend-segment-types git@github.com:qdrant/qdrant.git
```
TODO use `dev` once sparse vector API lands

agourlay commented on qdrant_client/http/models/models.py:
```diff
@@ -1880,15 +2133,22 @@ def __str__(self) -> str:
     List[StrictInt],
     List[StrictStr],
 ]
+# promoted at this place manually to avoid forward reference
```
Important workaround!

KShivendu commented on qdrant_client/http/api/cluster_api.py:
```diff
@@ -46,7 +46,7 @@ def _build_for_cluster_status(
         headers = {}
         return self.api_client.request(
-            type_=m.InlineResponse2002,
+            type_=m.InlineResponse2003,
```
```suggestion
            type_=m.InlineResponse203,
```

I think you wanted to do this. Or am I missing something?

agourlay commented:
async client generation issue due to https://github.com/qdrant/qdrant-client/issues/375

agourlay commented:
entering an early review process

agourlay commented on qdrant_client/local/async_qdrant_local.py:
```diff
@@ -596,19 +597,21 @@ async def create_collection(
         if src_collection and from_collection_name:
             batch_size = 100
-            records, next_offset = await self.scroll(
+            (records, next_offset) = await self.scroll(
```
No I did not try to workaround it, I have seen it will be fixed properly in https://github.com/qdrant/qdrant-client/pull/387

coszio replied:
Ah nice, didn't see that one 👍

agourlay commented:
It seems the payload selector feature did not work on the local mode.
I fixed it by porting the Rust implementation.

I also added doctests to CI https://github.com/qdrant/qdrant-client/pull/378/commits/319b5b82d5e8b93c2ec53b300e1fea54b69b947e

agourlay commented:
I have been pushing a lot of changes, time for a quick communication round :loudspeaker: 

## Current status

The local mode has been implemented and validated with congruence tests for:
- search
- retrieve
- scroll
- count

## Next

- add congruence tests for updates
- fix flaky search tests due to float precision
- fix async client generation
- get CI all green

coszio commented on tests/congruence_tests/test_common.py:
```diff
@@ -68,6 +79,30 @@ def generate_fixtures(
+) -> List[models.Record]:
+    if vectors_sizes is None:
+        vectors_sizes = {
+            "sparse-text": text_vector_size,
+            "sparse-image": image_vector_size,
+            "sparse-code": code_vector_size,
```
Does it not matter to use such small sizes?

agourlay replied:
Sorry I don't get it. Could you please elaborate? :)

coszio replied:
Sparse vectors are supposed to be used for big "dimensionality", but these are pretty small (<= 100). Is that alright for the tests?

agourlay commented on qdrant_client/local/local_collection.py:
```diff
@@ -347,15 +412,16 @@ def search(
         else:
             order = np.argsort(scores)
-
+        # TODO correct fix for https://github.com/qdrant/qdrant/pull/3082 ?
```
What I meant is this PR broke the Python client and the following line is the fix.
If you validate the fix and I will remove the TODO :)

coszio replied:
Ah, yes. That looks alright

agourlay commented on qdrant_client/local/local_collection.py:
```diff
@@ -315,29 +355,42 @@ def search(
+        else:
+            if name not in self.vectors:
+                raise ValueError(f"Dense vector {name} is not found in the collection")
+            else:
```
https://github.com/qdrant/qdrant-client/pull/378/commits/67e565999ebd44744997cdba38677c51acfdb63a

agourlay commented on qdrant_client/local/local_collection.py:
```diff
@@ -315,29 +355,42 @@ def search(
+            sort(query_vector)
+            if name in self.sparse_vectors:
+                vectors = self.sparse_vectors[name]
+                distance = Distance.DOT
```
https://github.com/qdrant/qdrant-client/pull/378/commits/67e565999ebd44744997cdba38677c51acfdb63a

agourlay commented on qdrant_client/local/local_collection.py:
```diff
@@ -901,20 +972,31 @@ def _update_point(self, point: models.PointStruct) -> None:
                 self.deleted_per_vector[vector_name][idx] = 1
 
+        # sparse vectors
+        for vector_name, named_vectors in self.sparse_vectors.items():
```
https://github.com/qdrant/qdrant-client/pull/378/commits/67e565999ebd44744997cdba38677c51acfdb63a

agourlay commented on tools/generate_grpc_client.sh:
```diff
@@ -6,7 +6,7 @@ PROJECT_ROOT="$(pwd)/$(dirname "$0")/../"
 cd $(mktemp -d)
 
-git clone --sparse --filter=blob:none --depth=1 -b dev git@github.com:qdrant/qdrant.git
+git clone --sparse --filter=blob:none --depth=1 -b sparse-index-try-to-extend-segment-types git@github.com:qdrant/qdrant.git
```
TODO use dev once sparse vector API lands

agourlay commented:
Need to check for a potential backward compatibility issue with the vector struct encoding for gRPC
